compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -802,6 +802,141 @@ static FailureOr<Flow::WorkloadBuilder> getWorkloadBuilder(OpBuilder &builder,
   return result;
 }
 
+/// Finds contiguous loop ranges from the given linalg.generic.
+static SmallVector<ReassociationIndices> getContiguousLoops(
+    linalg::GenericOp genericOp, SmallVectorImpl<unsigned> &loopDims) {
+  SmallVector<ReassociationIndices> contiguousLoops;
+  if (loopDims.size() < 2) return contiguousLoops;
+
+  auto isAllAffinemapsContiguous = [&](unsigned dim) {
+    return llvm::all_of(genericOp.getIndexingMapsArray(), [=](AffineMap map) {
+      return map.getDimPosition(dim) == dim;
+    });
+  };
+
+  ReassociationIndices range;
+  for (auto [idx, dim] : llvm::enumerate(loopDims)) {
+    if (!range.empty()) {
+      if (!isAllAffinemapsContiguous(dim) || (range.back() + 1 != dim)) {
+        if (range.size() > 1)
+          contiguousLoops.push_back({range.begin(), range.end()});
+        range.clear();
+      }
+      if (!range.empty() && !isAllAffinemapsContiguous(dim)) range.clear();
+    }
+    if (isAllAffinemapsContiguous(dim)) range.push_back(dim);
+  }
+  if (range.size() > 1) contiguousLoops.push_back(range);
+  return contiguousLoops;
+}
+
+/// Finds all collapsible loop ranges of type "parallel" and "reduction" without
+/// mixing.
+static SmallVector<ReassociationIndices> getCollapsibleLoops(
+    linalg::GenericOp genericOp) {
+  SmallVector<unsigned int> pDims, rDims;
+  genericOp.getParallelDims(pDims);
+
+  SmallVector<ReassociationIndices> pLoops =
+      getContiguousLoops(genericOp, pDims);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Collapsing parallel loops: ";
+    for (auto indices : pLoops) {
+      llvm::dbgs() << "[";
+      for (auto idx : indices) llvm::errs() << idx << ",";
+      llvm::dbgs() << "]\t";
+    }
+    llvm::dbgs() << "\n";
+  });
+
+  return pLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic and return the new
+/// one
+static FailureOr<linalg::GenericOp> collapseLinalgGeneric(
+    TensorDimTrackingRewriter &rewriter, linalg::GenericOp genericOp) {
+  SmallVector<ReassociationIndices> collapseIndices =
+      getCollapsibleLoops(genericOp);
+
+  rewriter.setInsertionPoint(genericOp);
+  Optional<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
```

Comment:
1. The return type is `FailureOr<SmallVector<Value>>`.

2. We should check failure with `failed(replacements)`, right?